### PR TITLE
Adds Q (from Q-criterion) calculation

### PR DESCRIPTION
**Pull request by tomchor**
This PR adds the calculation of the "Q" velocity gradient tensor invariant. This is useful for the identification and visualization of vortices in fluid dynamics (a comparative study of similar methods can be found [here](https://www.tandfonline.com/doi/pdf/10.1080/19942060.2020.1816496#:~:text=The%20Q%20criterion%20defines%20Q,strain%20in%20the%20vortices%20areas.&text=Lambda%2D2%20criterion%20(Jeong%20%26,%2B%202%20with%20%CF%832%20%3C%200).).)

The definition and nomenclature comes from the equation for the eigenvalues of the velocity gradient tensor $\nabla \vec u$:

```math
\lambda^3+P \lambda^2+Q \lambda+T=0
```

$Q$ is then defined as

```math
Q=\frac{1}{2}\left(\|\bf\Omega\|^2-\|\bf S\|^2\right)
```

where $\bf S$ is the strain rate tensor and $\bf \Omega$ is the vorticity tensor (i.e. the symmetric and antisymmetric parts of the velocity gradient tensor, respectively).

This PR adds `KernelFunctionOperations` for `StrainRateTensorModulus` ($|| \bf S ||$), `VorticityTensorModulus` ($|| \bf \Omega ||$),  and `Q`.

PS: I don't love this nomenclature since it gives basically zero information on what it's called, but I haven't ever seen it being called something else. If anyone knows of a better name for this, lmk.

### CODE REVIEW

File: src/FlowDiagnostics.jl [hunk 1]
```diff
@@ -315,4 +316,104 @@ function DirectionalErtelPotentialVorticity(model, direction; location = (Face,
...
+
+Its modulus is then defined (using Einstein summation notation) as
+
+    || Sᵢⱼ || = √( Sᵢⱼ Sᵢⱼ)
+
+"""
+function StrainRateTensorModulus(model; location = (Center, Center, Center))
+    validate_location(location, "StrainRateTensorModulus", (Center, Center, Center))
+    return KernelFunctionOperation{Center, Center, Center}(strain_rate_tensor_modulus_ccc, model.grid, model.velocities...)
+end
```
Comment (navidcy) [inline src/FlowDiagnostics.jl:350]: seems like kwarg `location` is not used?

File: src/FlowDiagnostics.jl [hunk 2]
```diff
@@ -315,4 +316,104 @@ function DirectionalErtelPotentialVorticity(model, direction; location = (Face,
...
+
+
+@inline fψ_minus_gφ²(i, j, k, grid, f, ψ, g, φ) = @inbounds (f(i, j, k, grid, ψ) - g(i, j, k, grid, φ))^2
```
Comment (navidcy) [inline src/FlowDiagnostics.jl:353]:
```suggestion
@inline fψ_minus_gφ²(i, j, k, grid, f, ψ, g, φ) = (f(i, j, k, grid, ψ) - g(i, j, k, grid, φ))^2
```

File: src/FlowDiagnostics.jl [hunk 3]
```diff
@@ -315,4 +316,104 @@ function DirectionalErtelPotentialVorticity(model, direction; location = (Face,
 end
 #---
 
+#+++ Velocity gradient tensor
+@inline fψ_plus_gφ²(i, j, k, grid, f, ψ, g, φ) = @inbounds (f(i, j, k, grid, ψ) + g(i, j, k, grid, φ))^2
```
Comment (navidcy) [inline src/FlowDiagnostics.jl:320]:
```suggestion
@inline fψ_plus_gφ²(i, j, k, grid, f, ψ, g, φ) = (f(i, j, k, grid, ψ) + g(i, j, k, grid, φ))^2
```

File: src/FlowDiagnostics.jl [hunk 4]
```diff
@@ -315,4 +316,104 @@ function DirectionalErtelPotentialVorticity(model, direction; location = (Face,
...
+
+Its modulus is then defined (using Einstein summation notation) as
+
+    || Ωᵢⱼ || = √( Ωᵢⱼ Ωᵢⱼ)
+
+"""
+function VorticityTensorModulus(model; location = (Center, Center, Center))
+    validate_location(location, "VorticityTensorModulus", (Center, Center, Center))
+    return KernelFunctionOperation{Center, Center, Center}(vorticity_tensor_modulus_ccc, model.grid, model.velocities...)
+end
```
Comment (navidcy) [inline src/FlowDiagnostics.jl:383]: similarly here, `location` kwarg is only validated but never used?

File: src/FlowDiagnostics.jl [hunk 5]
```diff
@@ -315,4 +316,104 @@ function DirectionalErtelPotentialVorticity(model, direction; location = (Face,
...
+
+    Q = ½ ( ΩᵢⱼΩᵢⱼ - SᵢⱼSᵢⱼ)
+
+and where `Sᵢⱼ= ½(∂ⱼuᵢ + ∂ᵢuⱼ)` and `Ωᵢⱼ= ½(∂ⱼuᵢ - ∂ᵢuⱼ)`. More info about it can be found in
+10.1063/1.5124245.
+"""
+function QVelocityGradientTensorInvariant(model; location = (Center, Center, Center))
+    validate_location(location, "QVelocityGradientTensorInvariant", (Center, Center, Center))
+    return KernelFunctionOperation{Center, Center, Center}(Q_velocity_gradient_tensor_invariant_ccc, model.grid, model.velocities...)
+end
```
Comment (navidcy) [inline src/FlowDiagnostics.jl:414]: same remark about `location`

**Review by navidcy — APPROVED** [review]

File: src/FlowDiagnostics.jl (hunk 2 above)
Comment (tomchor) [reply src/FlowDiagnostics.jl:353]: Just so that I understand, the reasoning here is that since there isn't a call to `getindex` here `@inbounds` is unnecessary?

Comment (navidcy) [reply src/FlowDiagnostics.jl:353]: Yes, `i, j, k` are just args in a function. There is no `getindex` involved.

File: src/FlowDiagnostics.jl (hunk 4 above)
Comment (tomchor) [reply src/FlowDiagnostics.jl:383]: Yes. The idea throughout these functions is to allow for a `location` argument. At the moment, however, since these have to implemented by hand using `Oceananigans.Operators` (and this being a new package with contributions mainly by myself), each diagnostic is only implemented in one location (most of them are at cell centers, although at least one (PV) is at cell edges). Hopefully, as things are needed and perhaps contributions come, there will be more location options for diagnostics.

So, in summary, that kwarg is there in anticipation of a future when we will have more options. One could certainly make the case to remove it, while adding superscripts in Oceananigans' style that indicates the location. So `VorticityTensorModulus` would become `VorticityTensorModulusᶜᶜᶜ` and the `location` wouldn't exist, but then I think the user interface suffers a bit.

If you have suggestions about other ways to manage this lmk! I'm currently not happy with either solution tbh

Comment (navidcy) [reply src/FlowDiagnostics.jl:383]: So you are saying that the problem is that `vorticity_tensor_modulus_ccc` needs to be redefine for each location, right?

Comment (tomchor) [reply src/FlowDiagnostics.jl:383]: Yes. `vorticity_tensor_modulus_ccf`, `vorticity_tensor_modulus_cfc` and so on...
